### PR TITLE
WIP: productionize filter code

### PR DESCRIPTION
**Pull request by osheroff**
@csenski @joshcanfield @zendesk/rules

### CODE REVIEW

File: src/main/java/com/zendesk/maxwell/Maxwell.java [hunk 1]
```diff
@@ -33,6 +32,28 @@ private void initFirstRun(Connection connection) throws SQLException, IOExceptio
 		this.context.setPosition(pos);
 	}
 
+	private void setupFilter(MaxwellFilter filter, MaxwellConfig config) {
+		if (!config.include_databases.isEmpty()) {
+			for ( String d : config.include_databases.split(",") )
```
Comment (joshcanfield) [inline src/main/java/com/zendesk/maxwell/Maxwell.java]: Is there value in supporting spaces in the string? 
`.split("\\s*,\\s*")`

e.g. `include_databases=one, two, three`

**Comment by osheroff** [comment]
@joshcanfield yeah just pushed latest with regexs & space handling and general better hygene, take a look.

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java [hunk 2]
```diff
@@ -1,57 +1,106 @@
...
-			return includeDatabases.contains(dbName);
+	private Pattern compile(String name) throws MaxwellInvalidFilterException {
+		if ( name.startsWith("/") ) {
+			if ( !name.endsWith("/") ) {
+				throw new MaxwellInvalidFilterException("Invalid regular expression: " + name);
+			}
+			return Pattern.compile(name.substring(1, name.length() - 1));
 		} else {
-			return !excludeDatabases.contains(dbName);
+			return Pattern.compile("^" + name.trim() + "$");
```
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: here is where we're handling leading/trailing spaces.  I guess we should trim the string before we start, could get regexs with leading whitespace...

File: src/main/java/com/zendesk/maxwell/MaxwellConfig.java [hunk 3]
```diff
@@ -60,11 +60,18 @@ private OptionParser getOptionParser() {
...
 		parser.accepts( "kafka.bootstrap.servers", "at least one kafka server, formatted as HOST:PORT[,HOST:PORT]" ).withRequiredArg();
 		parser.accepts( "kafka_topic", "optionally provide a topic name to push to. default: maxwell").withOptionalArg();
+
 		parser.accepts( "max_schemas", "how many old schema definitions maxwell should keep around.  default: 5").withOptionalArg();
 		parser.accepts( "init_position", "initial binlog position, given as BINLOG_FILE:POSITION").withRequiredArg();
 		parser.accepts( "replay", "replay mode, don't store any information to the server");
+
+		parser.accepts( "includeDatabases", "include thes databases, formatted as includeDatabases=db1,db2").withOptionalArg();
+		parser.accepts( "excludeDatabases", "exclude thes databases, formatted as excludeDatabases=db1,db2").withOptionalArg();
+		parser.accepts( "includeTables", "include thes tables, formatted as includeTables=db1,db2").withOptionalArg();
```
Comment (joshcanfield) [inline src/main/java/com/zendesk/maxwell/MaxwellConfig.java]: Typo - "thes" 
Should be updated with the regexp change too.

**Comment by joshcanfield** [comment]
Cool. That will be nice for excluding our history/audit tables.

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java [hunk 4]
```diff
@@ -1,57 +1,107 @@
...
-			return !excludeTables.contains(tblName);
+	public void addRowConstraint(String field, Integer value) {
+		rowFilter.put(field, value);
+	}
+
+	private boolean matchesIncludeExcludeList(List<Pattern> includeList, List<Pattern> excludeList, String name) {
+		if ( includeList.size() > 0 ) {
+			boolean found = false;
+			for ( Pattern p : includeList ) {
+				found = found || p.matcher(name).find();
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: break if found?

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java [hunk 5]
```diff
@@ -1,57 +1,107 @@
...
+			}
+			if ( !found )
+				return false;
+		}
+
+		for ( Pattern p : excludeList ) {
+			if ( p.matcher(name).find() )
+				return false;
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: Might want to add a note somewhere that excludes will override includes.

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java (hunk 2 above)
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: er. Is this even possible? Cmd line arguments themselves are separated with spaces. The more natural assumption would be that db/table names might have spaces in them (if mysql supports that), and assume users will provide db names as --include_databases='db name','another db'

Will the option parser strip out the quotes?

Comment (joshcanfield) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: Table names can also include commas...

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java (hunk 5 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: did an update pass at the configuration documentation.

File: src/main/java/com/zendesk/maxwell/MaxwellFilter.java (hunk 2 above)
Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: @vanchi-zendesk this code is here to handle the natural form: 

```--include_dbs="foo, bar"``` 

note that for spaces ```foo bar, baz` will work just fine.

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellFilter.java]: @joshcanfield it's true what you say about commas, but as I've told @vanchi-zendesk in other PRs, I refuse to code around people who act like sociopaths and put commas in their table names. :)

**Comment by osheroff** [comment]
@vanchi-zendesk I think I addressed all the CR concerns.

**Comment by vanchi-zendesk** [comment]
:+1:

File: docs/docs/config.md [hunk 6]
```diff
@@ -6,29 +6,55 @@ Maxwell can be configured by command line or a java "properties" file.
...
+Maxwell will send an update for a given row to its producer.  All the options take a single value PATTERN,
+which may either be a literal table/database name, given as `option=name`, or a regular expression,
+given as `option=/regex/`.  The options are evaluated as follows:
+
+1. only accept databases in `include_dbs`
+1. reject databases in `exclude_dbs`
+1. only accept tables in `include_tables`
+1. reject tables in `exclude_tables`
+
+So an example like `--include_dbs=/foo.*/ --exclude_tables=bar` will match `footy.baz` but not `footz.bar`.
```
Comment (joshcanfield) [inline docs/docs/config.md]: In the example the exclude isn't a regexp, so it will only match exactly bar, not 'footz.bar', right?

Comment (osheroff) [inline docs/docs/config.md]: in the example the database is `footz` (which matches the include) but the table is `bar`, so it will be excluded.

Comment (joshcanfield) [inline docs/docs/config.md]: Haha.. I blame Monday. Thanks.

File: docs/docs/config.md [hunk 7]
```diff
@@ -6,29 +6,55 @@ Maxwell can be configured by command line or a java "properties" file.
...
+Maxwell will send an update for a given row to its producer.  All the options take a single value PATTERN,
+which may either be a literal table/database name, given as `option=name`, or a regular expression,
+given as `option=/regex/`.  The options are evaluated as follows:
+
+1. only accept databases in `include_dbs` if non-empty
+1. reject databases in `exclude_dbs`
+1. only accept tables in `include_tables` if non-empty
+1. reject tables in `exclude_tables`
+
+So an example like `--include_dbs=/foo.*/ --exclude_tables=bar` will include `footy.zab` and exclude `footy.bar`
```
Comment (osheroff) [inline docs/docs/config.md]: rewrote the example a bit to be clearer, maybe?

The other thought I had was to try to boil it down to *only* include/exclude, so you could do:

```
 --include DB_NAME.*
 --include *.TBL_NAME
 --include /REGEX/ (where the regex is run against the fully qualified name.  you'd end up writing /.*\.prefix.*/ though )
```

whether that is really clearer though, I don't know.  It's a bit more flexible as you could tell maxwell to include tables `A.1` and `B.2` but *not* `A.2`, but feels slightly more cumbersome to work with.

I'm pretty sure Zendesk's filtering needs are handled entirely by `exclude_tables="/_.*_new/, /.*_stats$/"` but I could see the argument for more advanced cases.

**Pull request merged**